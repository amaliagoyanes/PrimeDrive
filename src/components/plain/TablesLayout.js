import React, {Component} from 'react';
import {
    Dimensions,
    TouchableOpacity,
    Text,
    View,
    ListView,
    Platform
} from 'react-native';
import Immersive from 'react-native-immersive';
import {Col, Row} from 'react-native-easy-grid';
import firestack from '../../fbconfig.js';
import {TableItem, ButtonItem} from '../index';
import {scale, scaleText} from '../../scaling';

import styles from '../../style/styles';

export default class TablesLayout extends Component {
    constructor(props) {
        super(props);

        this.state = {floors: [], tables: [], activeFloor: undefined};

        this.getFloors = this.getFloors.bind(this);
        this.getTables = this.getTables.bind(this);
    }

    getTableTypeButtonText(i) {
        switch (i) {
            case 0: return 'Free'; break;
            case 1: return 'Seated'; break;
            case 2: return 'Ordered'; break;
            case 3: return 'Starter'; break;
            case 4: return 'Main'; break;
            case 5: return 'Dessert'; break;
            case 6: return 'Check down'; break;
        }
    }

    getTableTypeButtonDetails(i) {
        switch (i) {
            case 0: return '12'; break;
            case 1: return '1'; break;
            case 2: return '1'; break;
            case 3: return '2'; break;
            case 4: return '3'; break;
            case 5: return '1'; break;
            case 6: return '1'; break;
        }
    }

    getTableStatusButtonBgColor(i) {
        switch (i) {
            case 'Free':
            case 0: return '#6f7076'; break;
            case 'Seated':
            case 1: return '#f52446'; break;
            case 'Ordered':
            case 2: return '#00d57d'; break;
            case 'Starter':
            case 3: return '#ffbb39'; break;
            case 'Main':
            case 4: return '#5083ff'; break;
            case 'Dessert':
            case 5: return '#ff59a3'; break;
            case 'Check down':
            case 6: return '#189bb5'; break;
        }
    }


    componentDidMount(){
        if (Platform.OS !== 'ios') {
            Immersive.on();
            Immersive.setImmersive(true);
        }
        this.getFloors();
    }

    componentDidUpdate(props, state) {
        if (state.activeFloor !== this.state.activeFloor)
            this.getTables();
    }

    getFloors() {
        if (this.state.floors !== undefined) {
            firestack.database.ref(`/floors/`).once('number').then((snapshot) => {
                let floors = snapshot.val();
                floors = Object.keys(floors).map(floor => {
                    return {...floors[floor], uid: floor}
                }).reverse();

                this.setState({floors: floors});
                console.log(floors)

                if (floors.length) {
                    this.setState({activeFloor: floors[0].uid});
                    this.getTables()
                }
            }).catch(err => console.log('failed get floors', err));
        }
    }

    getTables() {
        this.setState({tables: []});
        let extraFilter = this.props.extraFilter,
            orderBy = extraFilter ? 'floor_' + Object.keys(extraFilter).join('_'): 'floor',
            equalTo = extraFilter ? this.state.activeFloor + '~' + Object.values(extraFilter).join('~'): this.state.activeFloor;
        firestack.database.ref(`/tables/`).orderByChild(orderBy).equalTo(equalTo).once('name').then((snapshot) => {
            let tables = snapshot.val();
            if (tables) {
                tables = Object.keys(tables).map(table => {
                    return {...tables[table], uid: table}
                });
                if (tables.length)
                    this.setState({tables: tables});
            }
        }).catch(err => console.log('failed get tables', err));
    }

    render() {
        let perRow = 9,
            tablesLen = this.state.tables.length,
            rows = Math.ceil(tablesLen / perRow);
        return <Row size={20}>
            <Col>
                <Row size={1} style={{paddingLeft: scale(10)}}>
                    {this.state.floors.map((floor, i) =>
                        <ButtonItem key={i} textStyle={{fontSize: scaleText(8)}}
                                    onPress={() => this.setState({activeFloor: floor.uid})}
                                    text={'Floor ' + floor.number}
                                    active={floor.uid === this.state.activeFloor ? 'active' : '' }/>
                    )}
                </Row>
                <Row size={8} style={{paddingLeft: scale(10), paddingRight: scale(10)}}>
                    <Col style={styles.tablesLayout}>
                        {Array.apply(null, {length: rows}).map((d, idxRow) =>
                            <Row key={idxRow}>
                                {Array.apply(null, {length: perRow}).map((d, idxItem) => {
                                    let idx = idxRow * (perRow) + idxItem;
                                    if (idx < tablesLen) {
                                        let table = this.state.tables[idx];
                                        return <TableItem key={idxItem} num={table.name}
                                                          time={table.time ? table.time : undefined}
                                                          style={{backgroundColor: this.getTableStatusButtonBgColor(table.status)}}
                                                          onPress={() => this.props.onPress(table)}/>
                                    }
                                })}
                            </Row>
                        )}
                    </Col>
                </Row>
                {this.props.showTypeButtons &&
                <Row style={{paddingLeft: scale(10), paddingRight: scale(10)}}>
                    {Array.apply(null, {length: 7}).map((d, i) => {
                            return <ButtonItem key={i}
                                               style={[styles.buttonSidedText, {backgroundColor: this.getTableStatusButtonBgColor(i)}]}
                                               text={this.getTableTypeButtonText(i)}
                                               textStyle={{
                                                   fontSize: scaleText(8),
                                                   textAlign: 'left',
                                                   width: '50%'
                                               }}
                                               detailsText={this.getTableTypeButtonDetails(i)}/>
                        }
                    )}
                </Row>}
            </Col>
        </Row>
    }
}