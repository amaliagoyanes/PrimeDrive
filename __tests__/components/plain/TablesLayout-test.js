'use strict';

import 'react-native';
import React from 'react';
import {shallow} from 'enzyme';
import TablesLayout from '../../../src/components/plain/TablesLayout';

// Note: test renderer must be required after react-native.


let state = {
    activeFloor: 'floor_1',
    floors: [{uid: 'floor_1', number: 1}, {uid: 'floor_2', number: 2}, {uid: 'floor_3', number: 3}],
    tables: [
        {
            "uid": "table_1",
            "floor": "floor_1",
            "floor_status": "floor_1~Free",
            "name": "1",
            "status": "Free"
        },
        {
            "uid": "table_2",
            "floor": "floor_1",
            "floor_status": "floor_1~Starter",
            "name": "2nd",
            "status": "Starter"
        },
        {
            "uid": "table_3",
            "floor": "floor_2",
            "floor_status": "floor_2~Check down",
            "name": "3",
            "status": "Check down"
        },
        {
            "uid": "table_4",
            "floor": "floor_1",
            "floor_status": "floor_1~Free",
            "name": 4,
            "status": "Free"
        }
    ]

};

TablesLayout.prototype.getFloors = jest.fn;
TablesLayout.prototype.getTables = jest.fn;
TablesLayout.prototype.componentDidMount = jest.fn;

it('renders correctly', () => {
    const tree = shallow(
        <TablesLayout />
    );
    tree.setState(state);
    expect(tree).toMatchSnapshot();
});

it('renders correctly with type buttons', () => {
    const tree = shallow(
        <TablesLayout showTypeButtons={true}/>
    );
    tree.setState(state);
    expect(tree).toMatchSnapshot();
});
//TODO: extraFilter


