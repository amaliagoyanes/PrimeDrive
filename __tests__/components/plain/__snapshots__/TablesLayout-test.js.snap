// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RowNB
  size={20}
>
  <ColumnNB>
    <RowNB
      style={
        Object {
          "paddingLeft": 10,
        }
      }
    >
      <ButtonItem
        active="active"
        onPress={[Function]}
        text="Floor 1"
        textStyle={
          Object {
            "fontSize": 8,
          }
        }
      />
      <ButtonItem
        active=""
        onPress={[Function]}
        text="Floor 2"
        textStyle={
          Object {
            "fontSize": 8,
          }
        }
      />
      <ButtonItem
        active=""
        onPress={[Function]}
        text="Floor 3"
        textStyle={
          Object {
            "fontSize": 8,
          }
        }
      />
    </RowNB>
    <RowNB
      size={18}
      style={
        Object {
          "paddingLeft": 10,
          "paddingRight": 10,
        }
      }
    >
      <ColumnNB
        style={
          Object {
            "paddingBottom": 40,
            "paddingTop": 30,
          }
        }
      >
        <RowNB>
          <TableItem
            num="1"
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#6f7076",
              }
            }
          />
          <TableItem
            num="2nd"
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#ffbb39",
              }
            }
          />
          <TableItem
            num="3"
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#189bb5",
              }
            }
          />
          <TableItem
            num={4}
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#6f7076",
              }
            }
          />
        </RowNB>
      </ColumnNB>
    </RowNB>
  </ColumnNB>
</RowNB>
`;

exports[`renders correctly with type buttons 1`] = `
<RowNB
  size={20}
>
  <ColumnNB>
    <RowNB
      style={
        Object {
          "paddingLeft": 10,
        }
      }
    >
      <ButtonItem
        active="active"
        onPress={[Function]}
        text="Floor 1"
        textStyle={
          Object {
            "fontSize": 8,
          }
        }
      />
      <ButtonItem
        active=""
        onPress={[Function]}
        text="Floor 2"
        textStyle={
          Object {
            "fontSize": 8,
          }
        }
      />
      <ButtonItem
        active=""
        onPress={[Function]}
        text="Floor 3"
        textStyle={
          Object {
            "fontSize": 8,
          }
        }
      />
    </RowNB>
    <RowNB
      size={18}
      style={
        Object {
          "paddingLeft": 10,
          "paddingRight": 10,
        }
      }
    >
      <ColumnNB
        style={
          Object {
            "paddingBottom": 40,
            "paddingTop": 30,
          }
        }
      >
        <RowNB>
          <TableItem
            num="1"
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#6f7076",
              }
            }
          />
          <TableItem
            num="2nd"
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#ffbb39",
              }
            }
          />
          <TableItem
            num="3"
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#189bb5",
              }
            }
          />
          <TableItem
            num={4}
            onPress={[Function]}
            style={
              Object {
                "backgroundColor": "#6f7076",
              }
            }
          />
        </RowNB>
      </ColumnNB>
    </RowNB>
    <RowNB
      style={
        Object {
          "paddingLeft": 10,
          "paddingRight": 10,
        }
      }
    >
      <ButtonItem
        detailsText="12"
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "paddingLeft": 15,
              "paddingRight": 15,
            },
            Object {
              "backgroundColor": "#6f7076",
            },
          ]
        }
        text="Free"
        textStyle={
          Object {
            "fontSize": 8,
            "textAlign": "left",
            "width": "50%",
          }
        }
      />
      <ButtonItem
        detailsText="1"
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "paddingLeft": 15,
              "paddingRight": 15,
            },
            Object {
              "backgroundColor": "#f52446",
            },
          ]
        }
        text="Seated"
        textStyle={
          Object {
            "fontSize": 8,
            "textAlign": "left",
            "width": "50%",
          }
        }
      />
      <ButtonItem
        detailsText="1"
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "paddingLeft": 15,
              "paddingRight": 15,
            },
            Object {
              "backgroundColor": "#00d57d",
            },
          ]
        }
        text="Ordered"
        textStyle={
          Object {
            "fontSize": 8,
            "textAlign": "left",
            "width": "50%",
          }
        }
      />
      <ButtonItem
        detailsText="2"
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "paddingLeft": 15,
              "paddingRight": 15,
            },
            Object {
              "backgroundColor": "#ffbb39",
            },
          ]
        }
        text="Starter"
        textStyle={
          Object {
            "fontSize": 8,
            "textAlign": "left",
            "width": "50%",
          }
        }
      />
      <ButtonItem
        detailsText="3"
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "paddingLeft": 15,
              "paddingRight": 15,
            },
            Object {
              "backgroundColor": "#5083ff",
            },
          ]
        }
        text="Main"
        textStyle={
          Object {
            "fontSize": 8,
            "textAlign": "left",
            "width": "50%",
          }
        }
      />
      <ButtonItem
        detailsText="1"
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "paddingLeft": 15,
              "paddingRight": 15,
            },
            Object {
              "backgroundColor": "#ff59a3",
            },
          ]
        }
        text="Dessert"
        textStyle={
          Object {
            "fontSize": 8,
            "textAlign": "left",
            "width": "50%",
          }
        }
      />
      <ButtonItem
        detailsText="1"
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "paddingLeft": 15,
              "paddingRight": 15,
            },
            Object {
              "backgroundColor": "#189bb5",
            },
          ]
        }
        text="Check down"
        textStyle={
          Object {
            "fontSize": 8,
            "textAlign": "left",
            "width": "50%",
          }
        }
      />
    </RowNB>
  </ColumnNB>
</RowNB>
`;
